### PR TITLE
[Misc] Add a localStorage wrapper

### PR DESCRIPTION
**Pull request by Sindrake**
This is a spinoff from PR #586.

This PR adds a wrapper around localStorage, which makes handling values stored there easier.
Instead of setting or getting the value directly using a key, now you only need to call on the object property, and everything else will be handled automatically. This includes default values, type conversion, and caching if necessary.

For example, in order to get the current tag script, you previous would need to do this:
```
const current_script_id = LS.get("current_tag_script_id") || "1";
const tag_script = LS.get("tag-script-" + current_script_id) || "";
```
Now, it's just this:
```
const tag_script = Storage.Posts.TagScript.Content;
```

#### Note
Not all values were converted to the new format – some are using the legacy functions.
- The blacklist had been left untouched, since that whole system will be replaced in a different PR.
- Autocomplete does not really fit into this system, since the key seems to change.
- Theme switcher within `common.js` does not actually seem to do anything.

### CODE REVIEW

File: app/javascript/src/javascripts/utility/storage_util.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  for (const [accessor, params] of Object.entries(object))
+    StorageUtils.bootstrap(object, accessor, params[0], params[1]);
+};
+
+StorageUtils.getProxy = function (key, type, fallback) {
+  let val = localStorage.getItem(key);
+  if (val === null) return fallback;
+  switch (type) {
+    case "number":
+      return parseInt(val, 10);
```
Comment (DonovanDMC) [inline app/javascript/src/javascripts/utility/storage_util.js]:
```suggestion
      return Number(val);
```

File: app/javascript/src/javascripts/utility/storage.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  Set: ["set", 0],
+
+};
+StorageUtils.bootstrapMany(LStorage.Posts);
+
+LStorage.Posts.TagScript = {
+  /** @returns {number} Current tag script ID */
+  get ID () {
+    if (!this._tagScriptID)
+      this._tagScriptID = parseInt(localStorage.getItem("current_tag_script_id") || "1", 10);
```
Comment (DonovanDMC) [inline app/javascript/src/javascripts/utility/storage.js]:
```suggestion
      this._tagScriptID = Number(localStorage.getItem("current_tag_script_id") || "1");
```

File: app/javascript/src/javascripts/post_mode_menu.js [hunk 3]
```diff
@@ -30,29 +30,23 @@ PostModeMenu.show_notice = function (i) {
...
+  if ($("#mode-box-mode").val() !== "tag-script")
+    return;
+  e.preventDefault();
 
-    $("#tag-script-field").val(new_tag_script);
-    LS.put("current_tag_script_id", new_tag_script_id);
-    if (old_tag_script_id !== new_tag_script_id) {
-      PostModeMenu.show_notice(new_tag_script_id);
-    }
+  const newScriptID = parseInt(e.key, 10);
```
Comment (DonovanDMC) [inline app/javascript/src/javascripts/post_mode_menu.js]:
```suggestion
  const newScriptID = Number(e.key);
```

For base 10 `parseInt` and `Number` are roughly equivalent in functionality (assuming you don't need `parseInt`'s functionality of parsing up until the first non-number, `Number` bails out if any part of the string isn't a number), but `Number` is ~30% faster

**Review by DonovanDMC — CHANGES_REQUESTED** [review]

File: app/javascript/src/javascripts/post_mode_menu.js (hunk 3 above)
Comment (liamwhite) [reply app/javascript/src/javascripts/post_mode_menu.js]: The performance is irrelevant, right? It only matters that we only expect numbers as input, and so Number(x) is shorter than parseInt(x, 10) ...

Comment (DonovanDMC) [reply app/javascript/src/javascripts/post_mode_menu.js]: > The performance is irrelevant, right? It only matters that we only expect numbers as input, and so Number(x) is shorter than parseInt(x, 10) ...

On the (lack of) scale we're doing things, yes, the performance difference is largely irrelevant (the difference is measured in nanoseconds, which are billionths of a second)

Comment (Sindrake) [reply app/javascript/src/javascripts/post_mode_menu.js]: Keep in mind that using `Number` can result in some quirks.

```JS
parseInt("");    // returns NaN
parseInt(null);  // returns NaN
Number("");      // returns 0
Number(null);    // returns 0
```

I don't believe that this impacts us right now, but it is something to be aware of.

Comment (DonovanDMC) [reply app/javascript/src/javascripts/post_mode_menu.js]: > Keep in mind that using `Number` can result in some quirks.

Javascript is built on quirks, like `isNaN` vs `Number.isNaN`, it's great stuff

**Pull request merged**